djtriggers/models.py

```diff
@@ -82,31 +89,93 @@ def set_source(self, *args):
     def get_source(self):
         return tuple(x for x in self.source.split('$') if x != '')
 
-    def process(self, force=False, logger=None, dictionary={}):
-        if logger:
-            self.logger = get_logger(logger)
-        now = timezone.now()
-        if not force and self.date_processed is not None:
-            raise AlreadyProcessedError()
-        if not force and self.process_after and self.process_after >= now:
-            raise ProcessLaterError(self.process_after)
-
+    @shared_task(filter=task_method)
+    def process(self, force=False, logger=None, dictionary={}, use_statsd=False):
+        """
+        Executes the Trigger
+        :param bool force: force the execution
+        :param string logger: slug of preferred logger
+        :param dict dictionary: dictionary needed by trigger to execute
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        # The task gets locked because multiple tasks in the queue can process the same trigger.
+        # The lock assures no two tasks can process a trigger simultaneously.
+        # The check for date_processed assures a trigger is not executed multiple times.
         try:
-            self.logger.log_result(self, self._process(dictionary))
-        except ProcessLaterError as e:
-            self.process_after = e.process_after
-            self.save()
-            raise
-        if self.date_processed is None:
-            self.date_processed = now
-        self.save()
+            with NonBlockingLock.objects.acquire_lock(self):
+                if logger:
+                    self.logger = get_logger(logger)
+                now = timezone.now()
+                if not force and self.date_processed is not None:
+                    raise AlreadyProcessedError()
+                if not force and self.process_after and self.process_after >= now:
+                    raise ProcessLaterError(self.process_after)
+
+                try:
+                    # execute trigger
+                    self.logger.log_result(self, self._process(dictionary))
+                    self._handle_execution_success(use_statsd)
+                except ProcessLaterError as e:
+                    self.process_after = e.process_after
+                    self.save()
+                    raise
+                except Exception as e:
+                    self._handle_execution_failure(e, use_statsd)
+                    raise
+        except AlreadyLocked:
+            pass
```

Comment:
This task is run every X minutes, so if locked but didn't execute/complete, it will be retried by this job:
https://github.com/vikingco/django-triggers/blob/master/djtriggers/tasks.py#L7
Currently, the task is run with a management command, but we should convert that to a periodic task as described in the ticket.